Mathlib/RingTheory/Binomial.lean

```diff
@@ -368,6 +369,88 @@ theorem choose_natCast [NatPowAssoc R] (n k : ℕ) : choose (n : R) k = Nat.choo
+  rw [choose_zero_right', npow_zero]
+
+@[simp]
+theorem choose_zero_succ (R) [NonAssocRing R] [Pow R ℕ] [NatPowAssoc R] [BinomialRing R]
+    (n : ℕ) : choose (0 : R) (Nat.succ n) = 0 := by
+  unfold choose
+  rw [Nat.cast_succ, zero_sub, neg_add, neg_add_cancel_right, ← Nat.add_one,
+    multichoose_succ_neg_natCast]
+
+theorem choose_zero_pos (R) [NonAssocRing R] [Pow R ℕ] [NatPowAssoc R] [BinomialRing R]
+    {k : ℕ} (h_pos: 0 < k) : choose (0 : R) k = 0 := by
+  rw [← Nat.succ_pred_eq_of_pos h_pos, choose_zero_succ]
+
+theorem choose_zero_ite (R) [NonAssocRing R] [Pow R ℕ] [NatPowAssoc R] [BinomialRing R]
+    (k : ℕ) : choose (0 : R) k = if k = 0 then 1 else 0 := by
+  rw [eq_ite_iff]
+  by_cases hk: k = 0
+  constructor
+  rw [hk, choose_zero_right, ← Prod.mk.inj_iff]
+  right
+  constructor
+  exact hk
+  rw [← @Nat.le_zero, Nat.not_le] at hk
+  rw [choose_zero_pos R hk]
+
+@[simp]
+theorem choose_one_right' (r : R) : choose r 1 = r ^ 1 := by
+  rw [choose, Nat.cast_one, sub_add_cancel, multichoose_one_right']
+
+theorem choose_one_right [NatPowAssoc R] (r : R) : choose r 1 = r := by
+  rw [choose_one_right', npow_one]
+
+theorem descPochhammer_succ_succ_smeval {R} [NonAssocRing R] [Pow R ℕ] [NatPowAssoc R]
+    (r : R) (k : ℕ) : smeval (descPochhammer ℤ (Nat.succ k)) (r + 1) =
+    (k + 1) • smeval (descPochhammer ℤ k) r + smeval (descPochhammer ℤ (Nat.succ k)) r := by
+  nth_rw 1 [descPochhammer_succ_left]
+  rw [descPochhammer_succ_right, mul_comm (descPochhammer ℤ k)]
+  simp only [smeval_comp ℤ _ _ (r + 1), smeval_sub, smeval_add, smeval_mul, smeval_X, smeval_one,
+  npow_one, npow_zero, one_smul, add_sub_cancel_right, sub_mul, add_mul, add_smul, one_mul]
+  rw [← C_eq_natCast, smeval_C, npow_zero, add_comm (k • smeval (descPochhammer ℤ k) r) _,
+    add_assoc, add_comm (k • smeval (descPochhammer ℤ k) r) _, ← add_assoc,  ← add_sub_assoc,
+    nsmul_eq_mul, zsmul_one, Int.cast_natCast, sub_add_cancel, add_comm]
+
+theorem choose_succ_succ [NatPowAssoc R] (r:R) (k : ℕ) :
+    choose (r+1) (Nat.succ k) = choose r k + choose r (Nat.succ k) := by
+  refine nsmul_right_injective (Nat.factorial (k + 1)) (Nat.factorial_ne_zero (k + 1)) ?_
+  simp only [smul_add, ← descPochhammer_eq_factorial_smul_choose]
+  rw [Nat.factorial_succ, mul_smul,
+    ← descPochhammer_eq_factorial_smul_choose r, descPochhammer_succ_succ_smeval r k]
+
+theorem choose_eq_Nat_choose [NatPowAssoc R] (n k : ℕ) : choose (n : R) k = Nat.choose n k := by
+  induction n generalizing k with
+  | zero => cases k with
+    | zero => rw [choose_zero_right, Nat.choose_zero_right, Nat.cast_one]
+    | succ k => rw [Nat.cast_zero, choose_zero_succ, Nat.choose_zero_succ, Nat.cast_zero]
+  | succ n ih => cases k with
+    | zero => rw [choose_zero_right, Nat.choose_zero_right, Nat.cast_one]
+    | succ k => rw [Nat.cast_succ, choose_succ_succ, ih, ih, Nat.choose_succ_succ, Nat.cast_add]
+
+theorem choose_mul [NatPowAssoc R] (r : R) (n k : ℕ) (hkn : k ≤ n) :
```

Comment:
This should be `choose_smul`. (Possibly in fact `choose_smul_choose` but don't have a strong opinion.

Reply:
and below.